### ISSUE TITLE
Instance type recommendations

### ISSUE THREAD

**Issue by Glen-Moonpig** [npalm/terraform-aws-gitlab-runner#245]
I would be interested in recommendations for preferred instance types for different scenarios.

Up until now I have been using the general purpose (m4/m5) instances, currently 2xlarge, for compiling, testing and deployment jobs, but have recently found the spot capacity to be quite limited and the disk performance could be better.

I am now considering switching to c5.4xlarge for better CPUs and unrestricted EBS IO  but they come at a significantly higher cost. Or alternatively the SSD volumes of the i3.large may work better at a lower cost as long as the slower CPU clock speed doesn't make a significant difference.

Has anyone spent time optimising instance types and have any good recommendations in this area?

**Comment by kayman-mk** [comment]
We are using m4.xlarge instances to build Java projects and haven't found any problems so far. Also we didn't discover any problems to buy the spot instances.

For terraform jobs we do not use spot instances as the terraform state might be damaged if a terraform job is cancelled in between.

**Comment by Glen-Moonpig** [comment]
> We are using m4.xlarge instances to build Java projects and haven't found any problems so far. Also we didn't discover any problems to buy the spot instances.
> 
> For terraform jobs we do not use spot instances as the terraform state might be damaged if a terraform job is cancelled in between.

Thanks for you reply @kayman-mk. We used m4.xlarge for a while too but I think AWS were phasing them out in favor of the m5 types so we switched. We had a problem with a compatibility issue with m5 and NodeJS npx command which required a workaround.
We also have had some issue with terraform jobs being cancelled while running when spot capacity was low. We have had to build a self-service tool to remove the lock from the DynamoDb table when the job is cancelled before the lock is released because it was becoming a fairly common issue! Do you deploy this terraform module multiple times, with some using spot and some plain on-demand?

**Comment by kayman-mk** [comment]
@Glen-Moonpig Yeah, we deploy the module twice: terraform runner and non terraform runner. The first uses on demand with 0 idle instances and the latter uses spot instances with some put on idle.

This way we make sure not to damage the terraform state file.
